src/ripple/rpc/impl/KeypairForSignature.cpp

```diff
@@ -25,28 +25,102 @@ namespace RPC {
 
 KeyPair keypairForSignature (Json::Value const& params, Json::Value& error)
 {
-    if (! params.isMember ("secret"))
+    bool const has_key_type   = params.isMember (jss::key_type);
+    bool const has_passphrase = params.isMember (jss::passphrase);
+    bool const has_secret     = params.isMember (jss::secret);
+    bool const has_seed       = params.isMember (jss::seed);
+    bool const has_seed_hex   = params.isMember (jss::seed_hex);
+
+    int const n_secrets = has_passphrase + has_secret + has_seed + has_seed_hex;
+
+    if (n_secrets == 0)
     {
-        error = RPC::missing_field_error ("secret");
+        error = RPC::missing_field_error (jss::secret);
 
         return KeyPair();
     }
 
-    RippleAddress seed;
+    if (n_secrets > 1)
+    {
+        // `passphrase`, `secret`, `seed`, and `seed_hex` are mutually exclusive.
+        error = rpcError (rpcBAD_SECRET);
+
+        return KeyPair();
+    }
 
-    if (! seed.setSeedGeneric (params["secret"].asString ()))
+    if (has_key_type  &&  has_secret)
     {
-        error = RPC::make_error (rpcBAD_SEED,
-            RPC::invalid_field_message ("secret"));
+        // `secret` is deprecated.
+        error = rpcError (rpcBAD_SECRET);
 
         return KeyPair();
     }
 
+    KeyType type = KeyType::secp256k1;
+
+    RippleAddress seed;
+
+
+    if (has_key_type)
+    {
+        // `key_type` must be valid if present.
+
+        type = keyTypeFromString (params[jss::key_type].asString());
+
+        if (isInvalid (type))
+        {
+            error = rpcError (rpcBAD_SEED);
+
+            return KeyPair();
```

Comment:
`AnyPublicKey` and `AnySecretKey` constructors have a narrow contract: they all produce objects with valid states. `RippleAddress` offers no such assurances. In fact it has the `isValid` member to tell callers if its in a valid state.

Since this code is using the legacy `RippleAddress` with its share of additional problems, I'm inclined to just leave this alone. But it will be something to consider in future changes.